src/lib/storage/constraints/unique_checker.cpp

```diff
@@ -1,67 +1,51 @@
-    for (ChunkOffset chunk_offset = 0; chunk_offset < chunk->size(); chunk_offset++) {
-      const auto row_tid = mvcc_data->tids[chunk_offset].load();
-      const auto begin_cid = mvcc_data->begin_cids[chunk_offset];
-      const auto end_cid = mvcc_data->end_cids[chunk_offset];
-
-      auto row = TupleRow(constraint.columns.size());
-
-      if (Validate::is_row_visible(our_tid, snapshot_commit_id, row_tid, begin_cid, end_cid)) {
-        size_t row_index = 0;
-        for (const auto& column_id : constraint.columns) {
-          const auto& segment = segments[column_id];
-          const auto& value = (*segment)[chunk_offset];
-          // Since null values are considered unique (Assert(null != null)), we skip a row if we encounter a null value.
-          if (variant_is_null(value)) {
-            goto continue_with_next_row;
-          }
-          row[row_index++] = value;
-        }
-
-        const auto& [iterator, inserted] = unique_values.insert(row);
-        if (!inserted) {
-          return false;
-        }
-      }
-{continue_with_next_row:;}
-    }
-  }
-  return true;
+                          const CommitID snapshot_commit_id, const TransactionID our_tid) {
+  const auto checker = create_constraint_checker(table, constraint);
+  const auto& [satisfied, _] = checker->is_valid(snapshot_commit_id, our_tid);
+  return satisfied;
 }
 
-bool constraints_satisfied(const Table& table, const CommitID& snapshot_commit_id,
-                               const TransactionID& our_tid) {
-  for (const auto& constraint : table.get_unique_constraints()) {
-    if (!constraint_satisfied(table, constraint, snapshot_commit_id, our_tid)) {
-      return false;
+std::tuple<bool, ChunkID> check_constraints_for_values(const std::string& table_name,
+                                                           const std::shared_ptr<const Table>& table_to_insert,
+                                                           const CommitID snapshot_commit_id,
+                                                           const TransactionID our_tid, const ChunkID start_chunk_id) {
+  const auto table = StorageManager::get().get_table(table_name);
+  ChunkID first_chunk_to_check{0};
+  for (const auto& constraint : table->get_unique_constraints()) {
+    const auto checker = create_constraint_checker(*table, constraint);
+    const auto& [valid, chunk_id] = checker->is_valid_for_inserted_values(
+      table_to_insert, snapshot_commit_id, our_tid, start_chunk_id);
+    first_chunk_to_check = chunk_id;
+    if (!valid) {
+      return std::make_tuple<>(false, ChunkID{0});
     }
   }
-  return true;
-}
-
-bool constraints_satisfied(const std::string& table_name, const CommitID& snapshot_commit_id,
-                               const TransactionID& our_tid) {
-  auto const table = StorageManager::get().get_table(table_name);
-  return constraints_satisfied(*table, snapshot_commit_id, our_tid);
+  return std::make_tuple<>(true, first_chunk_to_check);
```

Comment:
Why do you include `first_chunk_to_check` in the return tuple? If it won't be used later, you could create a generic `constexpr` `SUCCESS_TUPLE` instead.

Reply:
When this function is called in the insert operator, this value is saved and then passed to the function as `start_chunk_id` during the commit phase (see `TransactionContext::commit_async`)